idelib/importer.py

```diff
@@ -354,14 +370,126 @@ def openFile(stream, updater=nullUpdater, parserTypes=elementParserTypes,
     return doc
 
 
-def readData(doc, source=None, updater=nullUpdater, numUpdates=500, updateInterval=.1,
+def filterTime(doc, startTime=0, endTime=None, channels=None):
+    """ Efficiently read data within a certain interval from an IDE file.
+        Note that due to the way data is stored in an IDE, the exported
+        interval will be slightly wider than the specified start and end
+        times; this ensures the data is copied verbatim and without loss.
+
+        :param doc: An opened (but not yet fully imported) `Dataset`.
+        :param startTime: The start of the extraction range, relative to the
+            recording's start.
+        :param endTime: The end of the extraction range, relative to the
+            recording's end.
+        :yields: Elements from the `Dataset`'s EBML file, excluding
+            `ChannelDataBlock`s outside of the specified time range, or
+            `None`.
+    """
+    if startTime == endTime:
+        raise ValueError('startTime and endTime must differ')
+
+    startTime, endTime = sorted((startTime or 0,
+                                 endTime or float('infinity')))
+
+    # Dictionaries (and similar) for tracking progress of each ChannelDataBlock
+    # element handled. All keyed by channel ID (ChannelIDRef).
+    lastBlocks = {}  # Previous element w/ its start and end times
+    channelsWritten = Counter()  # Number of elements extracted per channel
+    finished = {}  # Channels that have completed extraction
+
+    try:
+        timeScalars = doc._parsers['ChannelDataBlock'].timeScalars
+
+        for el in doc.ebmldoc:
+            if el.name == 'ChannelDataBlock':
+                # Get ChannelIDRef, StartTimeCodeAbs, and EndTimeCodeAbs;
+                # usually the 1st three, in order, but don't assume!
+                chId = blockStart = blockEnd = None
+                for subEl in el:
+                    if subEl.name == "ChannelIDRef":
+                        chId = subEl.value
+                    elif subEl.name == "StartTimeCodeAbs":
+                        blockStart = subEl.value
+                    elif subEl.name == "EndTimeCodeAbs":
+                        blockEnd = subEl.value
+
+                blockEnd = blockEnd or blockStart
```

Comment:
Is the `EndTimeCodeAbs` value missing from single-sample channels, or is it just allowed to be missing?

Reply:
It is allowed to be missing; this is what is done elsewhere in the Lab.